Fix cudaErrorInvalidDeviceFunction error caused by an uninstantiated functor

ldh4 opened the pull request:
To resolve #5474 

The error `cudaErrorInvalidDeviceFunction` is from `cudaFuncGetAttributes` when it's called during the deallocation of a dynamic ranked view. 
https://github.com/kokkos/kokkos/blob/dc7691829b6afcb25ffb4506a3a464b94cbc5078/core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp#L646-L659

The error indicates that the functor with `DestroyTag` was not instantiated at the time of `cudaFuncGetAttributes` call. This change is to instantiate that destroy functor during the construction stage, well before the destruction stage.

masterleinad reviewed (commented):
Hmmm.... We already have https://github.com/kokkos/kokkos/blob/26e9ba2bd5c670cac4f48a6befb82c6f721d167c/core/src/impl/Kokkos_ViewMapping.hpp#L3463-L3470 which should do the same thing. I wonder why that isn't sufficient.

dalg24 commented:
I was about to comment the same thing as Daniel.

Are you able to write a test that triggers that bug with the current `develop` branch?

ldh4 commented:
@masterleinad Yes, in fact, that's where I got the idea of trying out this change. My git bisect landed me at #5307, and then I tried applying the same workaround you already put in just to see if it helps suppressing this error. I agree that what was already there should have already done this instantiation.

@dalg24 I'll try dissecting the failing sacado unit test to see if I can come up with a Kokkos unit test for this. There is also a minimal reproducer mentioned in a [comment](https://github.com/kokkos/kokkos/issues/5474#issuecomment-1258826341) from a Trilinos issue that causes the same `cudaErrorInvalidDeviceFunction` from `cudaFuncGetAttributes`. However, seeing that this patch doesn't fix that minimal reproducer, that one may be a similar but a different issue.

crtrott commented:
@masterleinad the code snippet you show is not sufficient because that mapping specialiation is only used if the specialization tag is void, which isn't true for Stokhos/Sacado. My gut feeling is that this new change would be sufficient on its own, and we don't need the snipped you highlighted anymore. We should test that.

masterleinad commented:
> @masterleinad the code snippet you show is not sufficient because that mapping specialiation is only used if the specialization tag is void, which isn't true for Stokhos/Sacado. My gut feeling is that this new change would be sufficient on its own, and we don't need the snipped you highlighted anymore. We should test that.

That sounds plausible. 🙂

ldh4 commented:
@crtrott Unfortunately, it looks like this new change is also insufficient on its own, at least as it is right now. In a case where the mapping specialization tag is void and if a view is created with `Kokkos::WithoutInitializing` property, `construct_shared_allocation()` (the function where the changes in this PR are placed) is never called because of the following condition:
https://github.com/kokkos/kokkos/blob/e78d6a53a21da68a61cf1fd9dcefcb98fc161ad4/core/src/impl/Kokkos_ViewMapping.hpp#L3472-L3483
in which case the code snippet that was placed in by @masterleinad is necessary for the instantiation of the functor.

masterleinad commented:
@ldh4 Would pushing the workaround to the constructor of `ViewValueFunctor` be sufficient?

ldh4 commented:
@masterleinad Yep, that seems sufficient. Moved this workaround to the non-default constructors of non-scalar version of `ViewValueFunctor` and removed your workaround that was in `ViewMapping` class. This now seems to cover all cases.

masterleinad commented:
You need to fix the indentation. 🙂

masterleinad commented:
Retest this please.

ldh4 commented:
Retest this please.
